File: src/pipeline/chunking.ts
/**
 * Jagdhandbuch Merschbach — Pipeline Chunking Engine
 * 
 * Recursive Character Splitting für deutschsprachige Fachtexte
 * 
 * Best Practice 2025/2026 (Weaviate/Firecrawl):
 * - 400-512 Tokens (~2.000 Zeichen für Deutsch) pro Chunk
 * - 10-20% Overlap (200-400 Zeichen)
 * - Separator-Hierarchie: Absatz → Zeilenumbruch → Satz → Wort
 * 
 * Löst das 12.000-Zeichen-Problem: Statt 1 großen Chunk → 6+ kleine Chunks
 */

import * as stringSimilarity from 'string-similarity';

/**
 * Ein einzelner Chunk mit Metadaten
 */
export interface Chunk {
  id: string;
  text: string;
  index: number;
  start_char: number;
  end_char: number;
  token_estimate: number;
  source_document?: string;
  section_title?: string;
}

/**
 * Deduplizierungsergebnis
 */
export interface DeduplicationResult {
  original_count: number;
  deduplicated_count: number;
  removed_count: number;
  removed_facts: { fact: string; similar_to: string; similarity: number }[];
}

/**
 * Deutsche Satzgrenzen-Erkennung
 * Berücksichtigt: Abkürzungen (z.B., Nr., Abs., etc.), Zahlen, Aufzählungen
 */
const GERMAN_ABBREVIATIONS = new Set([
  'z.b', 'bzw', 'etc', 'usw', 'vgl', 'ggf', 'evtl', 'ca',
  'nr', 'abs', 'art', 'gem', 'lit', 'sog', 'inkl', 'exkl',
  'dr', 'prof', 'dipl', 'ing', 'min', 'max', 'std',
  'ha', 'km', 'kg', 'mg', 'ml', 'cm', 'mm',
  'jan', 'feb', 'mär', 'apr', 'mai', 'jun', 'jul', 'aug', 'sep', 'okt', 'nov', 'dez',
  'mo', 'di', 'mi', 'do', 'fr', 'sa', 'so',
  'str', 'pl', 'weg',
]);

function isSentenceBoundary(text: string, pos: number): boolean {
  if (pos >= text.length - 1) return true;
  
  const char = text[pos];
  if (char !== '.' && char !== '!' && char !== '?') return false;
  
  // Nach dem Punkt muss ein Leerzeichen und Großbuchstabe kommen
  const after = text.slice(pos + 1, pos + 3).trimStart();
  if (after.length === 0) return true;
  if (after[0] !== after[0].toUpperCase()) return false;
  
  // Prüfe ob vorher eine Abkürzung steht
  const before = text.slice(Math.max(0, pos - 10), pos).toLowerCase();
  const lastWord = before.split(/\s+/).pop() || '';
  if (GERMAN_ABBREVIATIONS.has(lastWord.replace(/\.$/, ''))) return false;
  
  // Prüfe auf Zahl vor dem Punkt (z.B. "3.5" oder "§ 12.")
  if (/\d$/.test(before) && /^\d/.test(after)) return false;
  
  return true;
}

/**
 * Teile Text an Satzgrenzen
 */
function splitIntoSentences(text: string): string[] {
  const sentences: string[] = [];
  let start = 0;
  
  for (let i = 0; i < text.length; i++) {
    if (isSentenceBoundary(text, i)) {
      const sentence = text.slice(start, i + 1).trim();
      if (sentence.length > 0) sentences.push(sentence);
      start = i + 1;
    }
  }
  
  // Rest
  const rest = text.slice(start).trim();
  if (rest.length > 0) sentences.push(rest);
  
  return sentences;
}

/**
 * Recursive Character Text Splitter für deutsche Fachtexte
 * 
 * Separator-Hierarchie (deutsch-optimiert):
 * 1. Doppelter Zeilenumbruch (Absatz)
 * 2. Einfacher Zeilenumbruch
 * 3. Satzende (. ! ?)
 * 4. Semikolon/Doppelpunkt
 * 5. Leerzeichen
 */
export function recursiveChunk(
  text: string,
  options: {
    /** Maximale Chunk-Größe in Zeichen (Default: 2000 ≈ 400-500 Tokens) */
    chunkSize?: number;
    /** Overlap in Zeichen (Default: 400 = 20% von 2000) */
    overlap?: number;
    /** Quell-Dokument Name */
    sourceDocument?: string;
  } = {}
): Chunk[] {
  const {
    chunkSize = 2000,
    overlap = 400,
    sourceDocument,
  } = options;
  
  const separators = ['\n\n', '\n', '. ', '! ', '? ', '; ', ': ', ' '];
  
  const chunks: Chunk[] = [];
  
  function splitRecursive(text: string, startChar: number, separatorIdx: number): void {
    // Wenn Text klein genug → direkt als Chunk
    if (text.length <= chunkSize) {
      if (text.trim().length > 0) {
        chunks.push({
          id: `chunk_${chunks.length}`,
          text: text.trim(),
          index: chunks.length,
          start_char: startChar,
          end_char: startChar + text.length,
          token_estimate: Math.ceil(text.length / 4.5), // ~4.5 chars/token für Deutsch
          source_document: sourceDocument,
        });
      }
      return;
    }
    
    // Kein Separator mehr → hart schneiden
    if (separatorIdx >= separators.length) {
      for (let i = 0; i < text.length; i += chunkSize - overlap) {
        const chunk = text.slice(i, i + chunkSize);
        if (chunk.trim().length > 0) {
          chunks.push({
            id: `chunk_${chunks.length}`,
            text: chunk.trim(),
            index: chunks.length,
            start_char: startChar + i,
            end_char: startChar + i + chunk.length,
            token_estimate: Math.ceil(chunk.length / 4.5),
            source_document: sourceDocument,
          });
        }
      }
      return;
    }
    
    // Versuche mit aktuellem Separator zu splitten
    const separator = separators[separatorIdx];
    const parts = text.split(separator);
    
    if (parts.length <= 1) {
      // Separator nicht gefunden → nächsten versuchen
      splitRecursive(text, startChar, separatorIdx + 1);
      return;
    }
    
    // Teile zusammenfügen bis chunkSize erreicht
    let currentChunk = '';
    let currentStart = startChar;
    
    for (const part of parts) {
      const candidate = currentChunk.length > 0 
        ? currentChunk + separator + part 
        : part;
      
      if (candidate.length > chunkSize && currentChunk.length > 0) {
        // Aktuellen Chunk speichern
        if (currentChunk.length <= chunkSize) {
          chunks.push({
            id: `chunk_${chunks.length}`,
            text: currentChunk.trim(),
            index: chunks.length,
            start_char: currentStart,
            end_char: currentStart + currentChunk.length,
            token_estimate: Math.ceil(currentChunk.length / 4.5),
            source_document: sourceDocument,
          });
        } else {
          // Zu groß → rekursiv mit nächstem Separator
          splitRecursive(currentChunk, currentStart, separatorIdx + 1);
        }
        
        // Overlap: Die letzten N Zeichen des vorherigen Chunks mitnehmen
        const overlapText = currentChunk.slice(-overlap);
        currentChunk = overlapText + separator + part;
        currentStart = currentStart + currentChunk.length - overlap - separator.length - part.length;
      } else {
        currentChunk = candidate;
      }
    }
    
    // Rest verarbeiten
    if (currentChunk.trim().length > 0) {
      if (currentChunk.length <= chunkSize) {
        chunks.push({
          id: `chunk_${chunks.length}`,
          text: currentChunk.trim(),
          index: chunks.length,
          start_char: currentStart,
          end_char: currentStart + currentChunk.length,
          token_estimate: Math.ceil(currentChunk.length / 4.5),
          source_document: sourceDocument,
        });
      } else {
        splitRecursive(currentChunk, currentStart, separatorIdx + 1);
      }
    }
  }
  
  splitRecursive(text, 0, 0);
  
  // Section Titles extrahieren (Markdown-Header)
  for (const chunk of chunks) {
    const headerMatch = chunk.text.match(/^#{1,3}\s+(.+)/m);
    if (headerMatch) {
      chunk.section_title = headerMatch[1].trim();
    }
  }
  
  return chunks;
}

/**
 * Fakt-Deduplizierung via String-Similarity
 * 
 * Threshold: 0.92 (Cosine Similarity)
 * Bei Überschreitung: Längeren Fakt behalten, kürzeren verwerfen
 */
export function deduplicateFacts(
  facts: string[],
  threshold: number = 0.92
): DeduplicationResult {
  const kept: string[] = [];
  const removed: { fact: string; similar_to: string; similarity: number }[] = [];
  
  for (const fact of facts) {
    let isDuplicate = false;
    
    for (const existing of kept) {
      const similarity = stringSimilarity.compareTwoStrings(
        fact.toLowerCase().trim(),
        existing.toLowerCase().trim()
      );
      
      if (similarity >= threshold) {
        isDuplicate = true;
        removed.push({ fact, similar_to: existing, similarity });
        
        // Wenn der neue Fakt länger ist, ersetze den bestehenden
        if (fact.length > existing.length) {
          const idx = kept.indexOf(existing);
          kept[idx] = fact;
          removed[removed.length - 1].similar_to = fact;
          removed[removed.length - 1].fact = existing;
        }
        break;
      }
    }
    
    if (!isDuplicate) {
      kept.push(fact);
    }
  }
  
  return {
    original_count: facts.length,
    deduplicated_count: kept.length,
    removed_count: removed.length,
    removed_facts: removed,
  };
}

File: pipeline/chunking.py
"""
Jagdhandbuch Merschbach — Pipeline Chunking Engine (Python)

Recursive Character Splitting für deutschsprachige Fachtexte.
Port der TypeScript-Implementierung (dashboard/src/pipeline/chunking.ts).

Best Practice 2025/2026:
- 2.000 Zeichen pro Chunk (~400-500 Tokens für Deutsch)
- 20% Overlap (400 Zeichen)
- Separator-Hierarchie: Absatz → Zeilenumbruch → Satz → Semikolon → Leerzeichen
- Deutsche Satzgrenzen-Erkennung (Abkürzungen, Zahlen, Aufzählungen)

Löst das 12.000-Zeichen-Problem:
  Vorher: 1 Chunk à 12.000 Zeichen → Dokument ab Seite 3 verloren
  Nachher: 6+ Chunks à 2.000 Zeichen → vollständige Extraktion
"""

from __future__ import annotations

import re
import hashlib
from dataclasses import dataclass, field
from typing import Optional

# ─────────────────────────────────────────────
# Deutsche Abkürzungen (Satzgrenzen-Erkennung)
# ─────────────────────────────────────────────

GERMAN_ABBREVIATIONS = frozenset([
    "z.b", "bzw", "etc", "usw", "vgl", "ggf", "evtl", "ca",
    "nr", "abs", "art", "gem", "lit", "sog", "inkl", "exkl",
    "dr", "prof", "dipl", "ing", "min", "max", "std",
    "ha", "km", "kg", "mg", "ml", "cm", "mm",
    "jan", "feb", "mär", "apr", "mai", "jun",
    "jul", "aug", "sep", "okt", "nov", "dez",
    "mo", "di", "mi", "do", "fr", "sa", "so",
    "str", "pl", "weg", "co", "gmbh", "ag", "e.v",
    "bzw", "zzgl", "inkl", "bzgl", "ggü", "i.d.r",
    "u.a", "o.ä", "d.h", "z.t", "s.o", "s.u",
])


def _is_sentence_boundary(text: str, pos: int) -> bool:
    """Prüfe ob Position eine Satzgrenze ist (deutsch-aware)."""
    if pos >= len(text) - 1:
        return True

    char = text[pos]
    if char not in ".!?":
        return False

    # Nach dem Satzzeichen muss Leerzeichen + Großbuchstabe kommen
    after = text[pos + 1:pos + 4].lstrip()
    if not after or not after[0].isupper():
        return False

    # Prüfe ob vorher eine Abkürzung steht
    before = text[max(0, pos - 15):pos].lower()
    words = before.split()
    if words:
        last_word = words[-1].rstrip(".")
        if last_word in GERMAN_ABBREVIATIONS:
            return False

    # Prüfe auf Zahl vor dem Punkt (z.B. "3.5" oder "§ 12.")
    if before and before[-1].isdigit() and after and after[0].isdigit():
        return False

    return True


# ─────────────────────────────────────────────
# Chunk Datenstruktur
# ─────────────────────────────────────────────

@dataclass
class Chunk:
    """Ein einzelner Text-Chunk mit Metadaten."""
    id: str
    text: str
    index: int
    start_char: int
    end_char: int
    token_estimate: int
    source_document: Optional[str] = None
    section_title: Optional[str] = None
    content_hash: str = ""

    def __post_init__(self):
        if not self.content_hash:
            self.content_hash = hashlib.sha256(
                self.text.encode("utf-8")
            ).hexdigest()[:16]


# ─────────────────────────────────────────────
# Recursive Character Text Splitter
# ─────────────────────────────────────────────

# Separator-Hierarchie (deutsch-optimiert)
SEPARATORS = ["\n\n", "\n", ". ", "! ", "? ", "; ", ": ", " "]


def recursive_chunk(
    text: str,
    chunk_size: int = 2000,
    overlap: int = 400,
    source_document: Optional[str] = None,
) -> list[Chunk]:
    """
    Teile einen Text rekursiv in Chunks auf.

    Args:
        text: Der zu teilende Text
        chunk_size: Maximale Chunk-Größe in Zeichen (Default: 2000 ≈ 400-500 Tokens)
        overlap: Overlap in Zeichen (Default: 400 = 20% von 2000)
        source_document: Name des Quelldokuments

    Returns:
        Liste von Chunk-Objekten
    """
    chunks: list[Chunk] = []

    def _split_recursive(txt: str, start_char: int, sep_idx: int) -> None:
        # Text passt in einen Chunk
        if len(txt) <= chunk_size:
            stripped = txt.strip()
            if stripped:
                chunks.append(Chunk(
                    id=f"chunk_{len(chunks):04d}",
                    text=stripped,
                    index=len(chunks),
                    start_char=start_char,
                    end_char=start_char + len(txt),
                    token_estimate=len(stripped) // 4,  # ~4 chars/token für Deutsch
                    source_document=source_document,
                ))
            return

        # Keine Separatoren mehr → hart schneiden
        if sep_idx >= len(SEPARATORS):
            for i in range(0, len(txt), chunk_size - overlap):
                piece = txt[i:i + chunk_size]
                stripped = piece.strip()
                if stripped:
                    chunks.append(Chunk(
                        id=f"chunk_{len(chunks):04d}",
                        text=stripped,
                        index=len(chunks),
                        start_char=start_char + i,
                        end_char=start_char + i + len(piece),
                        token_estimate=len(stripped) // 4,
                        source_document=source_document,
                    ))
            return

        separator = SEPARATORS[sep_idx]
        parts = txt.split(separator)

        if len(parts) <= 1:
            # Separator nicht gefunden → nächsten versuchen
            _split_recursive(txt, start_char, sep_idx + 1)
            return

        # Teile zusammenfügen bis chunk_size erreicht
        current_chunk = ""
        current_start = start_char

        for part in parts:
            candidate = (current_chunk + separator + part) if current_chunk else part

            if len(candidate) > chunk_size and current_chunk:
                # Aktuellen Chunk speichern oder rekursiv weiter splitten
                if len(current_chunk) <= chunk_size:
                    stripped = current_chunk.strip()
                    if stripped:
                        chunks.append(Chunk(
                            id=f"chunk_{len(chunks):04d}",
                            text=stripped,
                            index=len(chunks),
                            start_char=current_start,
                            end_char=current_start + len(current_chunk),
                            token_estimate=len(stripped) // 4,
                            source_document=source_document,
                        ))
                else:
                    _split_recursive(current_chunk, current_start, sep_idx + 1)

                # Overlap: letzte N Zeichen des vorherigen Chunks mitnehmen
                overlap_text = current_chunk[-overlap:] if overlap > 0 else ""
                current_chunk = overlap_text + separator + part
                current_start = (
                    current_start + len(current_chunk)
                    - len(overlap_text) - len(separator) - len(part)
                )
            else:
                current_chunk = candidate

        # Rest verarbeiten
        if current_chunk.strip():
            if len(current_chunk) <= chunk_size:
                stripped = current_chunk.strip()
                chunks.append(Chunk(
                    id=f"chunk_{len(chunks):04d}",
                    text=stripped,
                    index=len(chunks),
                    start_char=current_start,
                    end_char=current_start + len(current_chunk),
                    token_estimate=len(stripped) // 4,
                    source_document=source_document,
                ))
            else:
                _split_recursive(current_chunk, current_start, sep_idx + 1)

    _split_recursive(text, 0, 0)

    # Section Titles extrahieren (Markdown-Header)
    for chunk in chunks:
        header_match = re.search(r"^#{1,3}\s+(.+)", chunk.text, re.MULTILINE)
        if header_match:
            chunk.section_title = header_match.group(1).strip()

    return chunks


# ─────────────────────────────────────────────
# PDF-Extraktion + Chunking Pipeline
# ─────────────────────────────────────────────

def chunk_document(
    text: str,
    filename: str,
    chunk_size: int = 2000,
    overlap: int = 400,
) -> list[Chunk]:
    """
    Vollständige Dokument-Chunking-Pipeline.

    1. Text bereinigen (mehrfache Leerzeilen, Seitenumbrüche)
    2. Recursive Chunking
    3. Section Titles extrahieren
    4. Content Hashes berechnen

    Args:
        text: Extrahierter Dokumenttext
        filename: Dateiname des Quelldokuments
        chunk_size: Maximale Chunk-Größe
        overlap: Overlap-Größe

    Returns:
        Liste von Chunks mit Metadaten
    """
    # Bereinigung
    cleaned = re.sub(r"\n{3,}", "\n\n", text)       # Max 2 Leerzeilen
    cleaned = re.sub(r"\f", "\n\n", cleaned)          # Seitenumbrüche → Absatz
    cleaned = re.sub(r"[ \t]+\n", "\n", cleaned)      # Trailing Whitespace
    cleaned = re.sub(r"\n[ \t]+", "\n", cleaned)       # Leading Whitespace per Zeile
    cleaned = cleaned.strip()

    if not cleaned:
        return []

    return recursive_chunk(
        text=cleaned,
        chunk_size=chunk_size,
        overlap=overlap,
        source_document=filename,
    )
